### PR TITLE
Refactor chip-specific code into `esp_hal_common::soc` module

### PR DESCRIPTION
**Pull request by jessebraham**
The objective of this PR is to relocate as much chip-specific code as possible into this new `soc` module, much in the same way that ESP-IDF does. The intent behind this is to make our peripheral drivers a bit simpler while also making bringup of new devices a bit easier.

This does not seriously affect the public API of `esp-hal-common`, and mostly consists of internal changes.

Additional modules which can still be relocated:

- `esp_hal_common::clock::clocks_ll`
- `esp_hal_common::rtc_cntl::rtc`

I've opened this as a draft just to get some feedback on these changes. I believe this structure is better, however I want to make sure I'm not the only one who thinks so 😁 I'm happy to move the two modules listed above into `soc` in this PR as well, however they could also be moved in a separate PR; I'm not sure if there's anything else to move beyond that.

### CODE REVIEW

File: esp-hal-common/src/gpio.rs [hunk 1]
```diff
@@ -650,6 +630,16 @@ where
     PINTYPE: PinType,
     SIG: GpioSignal,
 {
+    pub fn new() -> Self {
```
Comment (MabezDev) [inline esp-hal-common/src/gpio.rs:633]: Remove this function, and use `steal()` instead wherever it is needed.

**Review by MabezDev — CHANGES_REQUESTED** [review]
This already looks so much better! Just one comment about the `new` function for GPIO.

File: esp-hal-common/src/gpio.rs (hunk 1 above)
Comment (jessebraham) [reply esp-hal-common/src/gpio.rs:633]: Ahh thanks will do! That was just copied from when I did this work previously, which I think predated `steal` 😁

**Review by bjoernQ — APPROVED** [review]
Looks much cleaner - good idea to have a `soc` module

I think the remaining things can get move in follow up PR(s) - no need to blow this up. But both is fine for me

**Review by JurajSadel — APPROVED** [review]
LGTM!

File: esp-hal-common/src/gpio.rs (hunk 1 above)
Comment (jessebraham) [reply esp-hal-common/src/gpio.rs]: `steal()` didn't exist so I've changed this to `pub(crate)` instead.

**Pull request merged**